### DISCUSSION TITLE
Build User Management App Magic Link not handling access_token

### DISCUSSION THREAD

**Discussion by taivo** [supabase/supabase#17716] [Questions]
Hello,
I'm following the "Build A User Management App With NextJs" tutorial here https://supabase.com/docs/guides/getting-started/tutorials/with-nextjs?language=ts and I'm not able to get the magic link authentication to work.

My setup uses the included inbucket setup. The email is properly sent and received. Clicking on the magic link does trigger  `auth/callback/route.ts`. However, because the access token is provided via the url hash, e.g. `#access_token=eyJhbGciOiJIUzI1NiIsInR....`, the server-side `route.ts` does not have access to that. From what I understand, when tokens are provided this way, the callback needs to be done on the client side.

What is the right thing to do here? I'd appreciate and pointers.

**Comment by sergchernata** [comment]
You should be able to parse that information out of the `req.url` object with a little string manipulation.

**Comment by taivo** [comment]
Hi @sergchernata , it doesn't look like the `#access_token` hash is passed in `req.url` to `auth/callback/route.ts`. I wasn't able to find it anywhere else in the `req` request object either. From what I've gathered so far, the hash part of urls aren't sent to the server.

**Reply by sergchernata** [reply]
If you're following the guide, and the guide doesn't have this issue, then you should double check that you have done everything correctly.

**Comment by porkopek** [comment]
I think that the guide works because they're using `Auth UI` from supabase, that handles magic link. But if you try to do it by yourself, with your own `<input />`, following the guide, it doesn't work

**Reply by sergchernata** [reply]
Right, the hash is available on the front, so that's probably how they're capturing it. OP is right, the hash isn't available server-side.

**Reply by porkopek** [reply]
The thing is if you send the magic link programatically, like
``` js
await supabase.auth.signInWithOtp({
      email:'email@gmail.com',
      options:{
        emailRedirectTo:'http://localhost:3000/auth/callback
      }})
 ```

the link you receive to click in your email has this searchParam `code`, that the route in auth/callback expects. But if you send the magic link in the dashboard through the button `send magic link`, the link you receive in your email has the hashproperty with "access_token" but no searchParams.
      I think it doesn't make sense. You should receive the same link. What do you think?

**Reply by sergchernata** [reply]
Agreed, that's a strange inconsistency. I'd create an issue for it.

**Reply by porkopek** [reply]
Further investigation, the magic link I receive when sending magic link from `signInWithOtp` has the following format (see bold)

jbsaczcvokxmtcirpnsh.supabase.co/auth/v1/verify?token=**pkce_xxx**&type=magiclink&redirect_to=http://localhost:3000/auth/callback

where the magic link sent through the dashboard has the following

jbsaczcvokxmtcirpnsh.supabase.co/auth/v1/verify?token=**xxx**&type=magiclink&redirect_to=http://localhost:3000/auth/callback`

The difference being that the link sent by `signInWithOtp` has this `pkce_` at the begining of the token, that might cause the app to redirect to a different url that when doesn't have this `pkce_` part

**Comment by porkopek** [comment]
@taivo 
I think this is the current way to handle it now
https://supabase.com/docs/guides/auth/server-side/email-based-auth-with-pkce-flow-for-ssr.

A new route 
```js 
export async function GET(req) {
  const { searchParams } = new URL(req.url)
  const token_hash = searchParams.get('token_hash')
  const type = searchParams.get('type')
  const next = searchParams.get('next') ?? '/'

  if (token_hash && type) {
    const supabase = createRouteHandlerClient({ cookies })
    const { error } = await supabase.auth.verifyOtp({ type, token_hash })
    if (!error) {
      return NextResponse.redirect(new URL(`/${next.slice(1)}`, req.url))
    }
  }

  // return the user to an error page with some instructions
  return NextResponse.redirect(new URL('/auth/auth-code-error', req.url))
}
```

 I tried and it works both for supabase dashboard and supabase next auth helpers

**Comment by taivo** [comment]
I was using the `Auth UI` as described in that tutorial when this issue happened. I think this comes down to the difference between cloud supabase and the local dev supabase setup. Here are the relevant steps that lead to this:
1. `pnpm dev` to run the nextjs server on localhost:3000
2. click the send magic link provided by Auth UI on localhost:3000
3. Open another browser tab to see the local inbucket server on http://localhost:54324 and click on the magic link
4. The issue described above happens, console.log within `auth/callback/route.ts` shows that `req.url` is simply `http://localhost:3000/auth/callback` with no hash or search parameter

The magic link received by inbucket looks like this:
http://localhost:54321/auth/v1/verify?token=xxx&type=magiclink&redirect_to=http://localhost:3000/auth/callback

So it looks like the api server on port 54321 is handling the code exchange then redirect to http://localhost:3000/auth/callback#access_code=xxx 

This is the type of url that can easily be parsed by frontend code. However, the hash is not passed to the backend so there is nothing for `auth/callback/route.ts` to process. This is a tad complicated because the round trip involves servers on 3 different ports (nextJs 3000, in bucket 54324, and supabase api 54321). I think I'll have to read up a bit more to see what this api server is all about.

I was initially hoping to just use the Auth UI instead of building my own auth forms (because I planned on adding social providers as well). 

@porkopek thanks for the new route implementation. I'll put it to use with the auth-helpers if I can't get this Auth UI to work.
